### PR TITLE
Add support for float8_e4m3fnuz and _e5m2fnuz

### PR DESCRIPTION
**Pull request by galexite**
This PR relates to the feature in [this feature submission](https://docs.google.com/document/d/1pF2T1xz54IPg1jG7FhykbrpbcJZVelQw0v8vBaoLkfs/edit). It has been based on #104242 which adds similar float8 types.

These new types added in this PR are described in the paper at https://arxiv.org/abs/2206.02915. A brief description and comparison of the types with other float8 types can be also found in the [OpenXLA RFC](https://github.com/openxla/stablehlo/blob/main/rfcs/20230321-fp8_fnuz.md).

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang @albanD

### CODE REVIEW

**Comment by galexite** [comment]
@albanD @seemethere – would it be possible to get an exception for the 2000 LOC limit? I'm only 6 lines over and quite a lot of it is just registering the types.

**Comment by galexite** [comment]
I've removed my changes to `finfo` to fit inside the LOC limit – I'll commit them as a separate PR.

**Comment by vkuzo** [comment]
Thanks for working on this! At a high level this looks good, and we will need to do a more detailed review.  Given that the branch cut for v2.1 is only a few days away and the high LOC of this change and our experience with landing previous "new dtype" PRs, I would expect that Pytorch v2.2 would be a reasonable target for eventually getting this in.

File: c10/util/Float8_e4m3fnuz.h [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+namespace detail {
+
+/*
+ * Convert a 8-bit floating-point number in fp8 E4M3FNUZ format, in bit
+ * representation, to a 32-bit floating-point number in IEEE single-precision
+ * format, in bit representation.
+ *
+ * @note The implementation doesn't use any floating-point operations.
+ */
+inline C10_HOST_DEVICE float fp8e4m3fnuz_to_fp32_value(uint8_t input) {
```
Comment (vkuzo) [inline c10/util/Float8_e4m3fnuz.h:63]: this looks reasonable.  Just curious, what made you choose a LUT over bit shifting?

also, do we expect the hardware to support an accelerated version of these?

File: test/quantization/core/experimental/test_float8.py [hunk 2]
```diff
@@ -91,7 +128,15 @@ class TestFloat8Dtype(TestCase):
     Sanity test for zeros comparison
     """
 
-    @parametrize("dtype", [torch.float8_e5m2, torch.float8_e4m3fn])
+    @parametrize(
+        "dtype",
```
Comment (vkuzo) [inline test/quantization/core/experimental/test_float8.py:132]: nit: maybe something like

```
FLOAT8_DTYPES = [...]

@parametrize("dtype", FLOAT8_DTYPES)
```

to save some repetitive LOC?

**Comment by vkuzo** [comment]
The dtype pieces looks reasonable to me!  Wondering if you could share some context on which hardware supports these float8 flavors now, and which hardware is expected to support this in the future?

cc @malfet , would you be up for a more detailed review on the framework pieces figuring out the best way to land this?

File: c10/util/Float8_e4m3fnuz.h (hunk 1 above)
Comment (galexite) [reply c10/util/Float8_e4m3fnuz.h:63]: There's no particular reason to use a LUT here, I can change to bit shifting if needed!

Yep, Graphcore's C600 hardware has dedicated instructions for which can be used to convert to and from both these FP8 types.

**Comment by galexite** [comment]
Thanks for the review @vkuzo!

> The dtype pieces looks reasonable to me! Wondering if you could share some context on which hardware supports these float8 flavors now, and which hardware is expected to support this in the future?

Graphcore's [C600 card](https://www.graphcore.ai/products/c600) support these types at the hardware level and these currently available on the market. It has instructions in the [Tile ISA](https://docs.graphcore.ai/projects/isa-mk2-with-fp8/en/latest/_static/TileVertexISA-IPU21-1.3.1.pdf) to perform common operations directly on FP8 data, as well as convert between types.

I'll make the change to take those list of dtypes for the test parameters out in to a constant.

**Comment by galexite** [comment]
Yeah, it looks like it has gone over. If that's not acceptable, I'm happy to remove it and submit separately.

**Comment by galexite** [comment]
@albanD, @seemethere: would it be possible to make an exception for this diff? I'm 42 lines over – I felt leaving the `torch.finfo` support in makes sense as the information required is already in the diff.

**Review by seemethere — APPROVED** [review]
LGTM

**Comment by galexite** [comment]
@malfet, may I trouble you for a review please? Thanks

**Review by malfet — CHANGES_REQUESTED** [review]
This PR feels very much like a copy-n-paste of https://github.com/pytorch/pytorch/pull/104242 that adds 2 more very similar data types.
Which looks fine on the surface, but IMO duplicates the code for F8 data types and leaves lots of questions unanswered, for example how how dtype conversion should work between two f8_e5m2 flavors.

Perhaps more importantly, what is the goal of this PR?  https://github.com/pytorch/pytorch/pull/104242 was a stepping stone for enabling F8 support in ARM cpus and NVIDIA H100 GPUs, but it looks like NZ flavors are only available in GraphCore AI accelerators. If this is the case, then perhaps it would be sufficient to just use `torch.bits8` dtype or indeed add a new enum, but skip dispatch implementation for the basic OPs?

Also, would it be possible to share some of the implementation between say F8_E5M2 and F8_E5M2NZ?

File: aten/src/ATen/AccumulateType.h [hunk 3]
```diff
@@ -106,7 +108,9 @@ CUDA_ACC_TYPE(half, float);
 CUDA_ACC_TYPE(BFloat16, float);
 CUDA_ACC_TYPE(Half, float);
 CUDA_ACC_TYPE(Float8_e5m2, float);
+CUDA_ACC_TYPE(Float8_e5m2fnuz, float);
```
[the remark this answers is not in the document]
Comment (galexite) [reply aten/src/ATen/AccumulateType.h:111]: Both of these were a bad merge on my part – they are not necessary. I've removed them.

**Comment by galexite** [comment]
Thanks for the review @malfet!

> This PR feels very much like a copy-n-paste of #104242 that adds 2 more very similar data types. Which looks fine on the surface, but IMO duplicates the code for F8 data types

Which code in particular? The operator definitions defined in `Float8_e{5m2,4m3}fnuz-inl.h` especially are quite large in LOC and could easily be covered by macros shared between all the float8 types, but they were already duplicated between `Float8_e5m2-inl.h` and `Float8_e4m3fn-inl.h`, so I chose to do the same.

> and leaves lots of questions unanswered, for example how how dtype conversion should work between two f8_e5m2 flavors.

I'm not sure how common it would be to convert between these dtypes, especially if hardware only supports one or the other – I'm happy to provide casts however.

> Perhaps more importantly, what is the goal of this PR? #104242 was a stepping stone for enabling F8 support in ARM cpus and NVIDIA H100 GPUs, but it looks like NZ flavors are only available in GraphCore AI accelerators. If this is the case, then perhaps it would be sufficient to just use `torch.bits8` dtype or indeed add a new enum, but skip dispatch implementation for the basic OPs?

The goal of this PR is to have native support for fnuz types as supported by hardware inside PyTorch. Ideally, this would involve some support for basic ops on CPU (add, mul, copy_, ne, etc.) in the same manner as they are for e5m2, e4m3.

> but it looks like NZ flavors are only available in GraphCore AI accelerators

Support for these types [have been added to the AMDGPU backend of LLVM](https://github.com/llvm/llvm-project/commit/22f0c7a45149a55e643665c4cb69faa960e3c565) for native hardware support on AMD's `gfx940` architecture.

> Also, would it be possible to share some of the implementation between say F8_E5M2 and F8_E5M2NZ?

I'm not sure what you mean by 'share some of the implementation between' – the C600 does not support e5m2, only e5m2fnuz, likewise for e4m2.

**Comment by galexite** [comment]
@malfet: I hope I've responded to your questions well enough! Is there a chance you could re-review?

**Comment by jeffdaily** [comment]
> @malfet: I hope I've responded to your questions well enough! Is there a chance you could re-review?

@malfet I too request a re-review.  As pointed out in prior comments, these types are necessary for our gfx940.

**Comment by malfet** [comment]
@jeffdaily , @galexite this PR fails 16+ tests because it's trying to run something with CUDA using new dtypes, which:
- Renders it unmergeable
- Not something we want to run/test/compile for, don't we?

File: aten/src/ATen/AccumulateType.cpp [hunk 4]
```diff
@@ -15,7 +15,7 @@ c10::ScalarType toAccumulateType(c10::ScalarType type, c10::DeviceType device) {
           return CppTypeToScalarType<at::acc_type_device<scalar_t, c10::DeviceType::CPU>>::value;  \
       }
 
-    AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF(DEFINE_CASE)
+    AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF_F8NZ(DEFINE_CASE)
```
Comment (malfet) [inline aten/src/ATen/AccumulateType.cpp:18]: Can you please explain why this one is needed?

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 5]
```diff
@@ -73,7 +73,9 @@ struct is_floating_point:
       std::is_same<T, at::Half>::value ||
       std::is_same<T, at::BFloat16>::value ||
       std::is_same<T, at::Float8_e5m2>::value ||
-      std::is_same<T, at::Float8_e4m3fn>::value> {
+      std::is_same<T, at::Float8_e5m2fnuz>::value ||
+      std::is_same<T, at::Float8_e4m3fn>::value ||
+      std::is_same<T, at::Float8_e4m3fnuz>::value> {
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec_base.h:78]: I think it's safe to remove all float8 dtypes from here for now, as there shouln't be any vec256 instantiations (I can do it in a separate PR though)

File: aten/src/ATen/native/CPUBlas.cpp [hunk 6]
```diff
@@ -581,7 +581,7 @@ void gemm_batched_with_stride(
       scalar_t beta,                                            \
       scalar_t *c, int64_t ldc, int64_t batch_stride_c);
 
-AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF(INSTANTIATE_BATCHED_GEMM)
+AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF_F8NZ(INSTANTIATE_BATCHED_GEMM)
```
Comment (malfet) [inline aten/src/ATen/native/CPUBlas.cpp:584]: Again curious, why this change is needed here?

File: aten/src/ATen/native/cuda/CompareEQKernel.cu [hunk 7]
```diff
@@ -28,12 +28,22 @@ struct CompareEqFunctor{
...
+      kFloat8_e4m3fn,
+      kFloat8_e4m3fnuz,
+      kFloat8_e5m2,
+      kFloat8_e5m2fnuz,
+      iter.common_dtype(),
+      "compare_eq_ne_cuda",
+      [&]() {
+        opmath_symmetric_gpu_kernel_with_scalars<scalar_t, bool>(
+            iter, CompareEqFunctor<scalar_t>(op));
+      });
```
Comment (malfet) [inline aten/src/ATen/native/cuda/CompareEQKernel.cu:46]: If you don't mind, let's leave those until a followup PRs?

File: aten/src/ATen/native/cuda/Copy.cu [hunk 8]
```diff
@@ -107,8 +107,8 @@ void direct_copy_kernel_cuda(TensorIteratorBase &iter) {
   } else if (dtype == kFloat8_e5m2 || dtype == kFloat8_e4m3fn) {
      float8_copy_kernel_cuda(iter);
   } else {
-    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
-        kHalf, kBool, kBFloat16, kComplexHalf,dtype, "copy_", [&] {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(
+        kHalf, kBool, kBFloat16, kComplexHalf, kFloat8_e5m2fnuz, kFloat8_e4m3fnuz, dtype, "copy_", [&] {
```
Comment (malfet) [inline aten/src/ATen/native/cuda/Copy.cu:111]: Same as above, let's leave it until followups

**Comment by jeffdaily** [comment]
To fix the CI failures, is it enough to update aten/src/ATen/native/cuda/FillKernel.cu and add the 2 new dispatch types?

File: c10/util/Float8_e4m3fnuz.h (hunk 1 above)
Comment (malfet) [reply c10/util/Float8_e4m3fnuz.h:63]: Did we run any perf benchmarks/binary size consideration for having this lookup table embedded over and over in every op that will need to convert F8E4M3FNUZ to float?

**Comment by galexite** [comment]
Sorry @malfet, @jeffdaily, I backed up some incomplete changes to this PR and have been working in my local copy since. I have uploaded the rest which removes all the CUDA kernels and disables the test for CUDA. The tests will still fail, because the `simulateFp8Precision` function does not work for a certain input for e4m3fnuz, and I haven't got round to debugging it yet.

File: aten/src/ATen/AccumulateType.cpp (hunk 4 above)
Comment (galexite) [reply aten/src/ATen/AccumulateType.cpp:18]: I changed the name of the `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF` macro to make it clear it isn't all scalar types. Unfortunately, were I to add F8NZ types to that macro, I'd have to add CUDA_ACC_TYPEs and MPS_ACC_TYPEs for these.

**Comment by jeffdaily** [comment]
@galexite thanks.  CI is now passing.  The only remaining issue is responding to @malfet's question about the LUT possibly bloating the binary size.

**Comment by jeffdaily** [comment]
@galexite have you had a chance to respond to @malfet's question?

**Comment by AnthonyBarbier** [comment]
> @galexite have you had a chance to respond to @malfet's question?

Sorry, he's away, he'll be back on Monday. Which question are you talking about? (Can't see any unanswered thread?)

**Comment by jeffdaily** [comment]
> > @galexite have you had a chance to respond to @malfet's question?
> 
> Sorry, he's away, he'll be back on Monday. Which question are you talking about? (Can't see any unanswered thread?)

There was concern that the use of the lookup table in the header file would bloat binary sizes.  So @galexite said they would move it from the header to the cpp file.

**Comment by galexite** [comment]
I moved `c10::detail::fp8{e4m3,e5m2}fnuz_to_fp32_value` to their respective .cpp files, and have exported them using `C10_API`.

**Comment by galexite** [comment]
The issue with the test is that `simulateFp8Precision` does not properly round values which would be subnormal in the float8 type to the nearest valid subnormal. This is even the case for the existing types, but it appears troublesome values do not get generated in `x` for the test to fail for these existing types. I have added an extra test for subnormals and have been trying to figure out how to simulate them properly.

As an aside, the `assertEqual` should probably have both tolerances set to 0.

**Comment by jeffdaily** [comment]
@galexite build failures are real.  Missing symbol for both rocm and cuda builds.

`lld: error: undefined symbol: c10::detail::fp8e4m3fnuz_to_fp32_value(unsigned char)`

**Comment by galexite** [comment]
> @galexite build failures are real. Missing symbol for both rocm and cuda builds.
> 
> `lld: error: undefined symbol: c10::detail::fp8e4m3fnuz_to_fp32_value(unsigned char)`

Sorry @jeffdaily, I should have posted here. I'm looking in to that today but we had to re-write the tests (and add a load more), which I've pushed up as a separate PR here: https://github.com/pytorch/pytorch/pull/113361. I'll rebase this PR atop that other one when it lands.

**Comment by galexite** [comment]
I think I have worked around the linking issue with CUDA and ROCm, there is just an issue with overflows from the e5m2fnuz cast which I need to resolve (those are the test failures just appearing now).

File: c10/util/Float8_e5m2fnuz-inl.h [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      numeric_limits<float>::tinyness_before;
+
+  static constexpr c10::Float8_e5m2fnuz min() {
+    return c10::Float8_e5m2fnuz(0x4, c10::Float8_e5m2fnuz::from_bits());
+  }
+  static constexpr c10::Float8_e5m2fnuz max() {
+    return c10::Float8_e5m2fnuz(0x7B, c10::Float8_e5m2fnuz::from_bits());
+  }
+  static constexpr c10::Float8_e5m2fnuz lowest() {
+    return c10::Float8_e5m2fnuz(0xFB, c10::Float8_e5m2fnuz::from_bits());
```
Comment (umangyadav) [inline c10/util/Float8_e5m2fnuz-inl.h:79]: This violates CPP standard which says `max() == -lowest()` 
`lowest` should return minimum value representable such that there is not `x` which is smaller than lowest ` x< lowest()` 

This should return `0xFF`

File: c10/util/Float8_e5m2fnuz-inl.h [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  static constexpr int max_exponent10 = 4;
+  static constexpr auto traps = numeric_limits<float>::traps;
+  static constexpr auto tinyness_before =
+      numeric_limits<float>::tinyness_before;
+
+  static constexpr c10::Float8_e5m2fnuz min() {
+    return c10::Float8_e5m2fnuz(0x4, c10::Float8_e5m2fnuz::from_bits());
+  }
+  static constexpr c10::Float8_e5m2fnuz max() {
+    return c10::Float8_e5m2fnuz(0x7B, c10::Float8_e5m2fnuz::from_bits());
```
Comment (umangyadav) [inline c10/util/Float8_e5m2fnuz-inl.h:76]: This should be `0x7F` for the FNUZ types. FNUZ types do not have  infinity. 
NaN is represented by negative zero. So `7C`, `7D`, 7E` and `7F` are available in representable range.

**Comment by galexite** [comment]
Yeah, there were a couple of issues with the `numeric_limits`, I hope I've fixed all of them.

**Review by malfet — APPROVED** [review]
Thank you very much for a nice refactors, looks good to me.

File: c10/util/Float8_e4m3fnuz.h [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+ * Convert a 32-bit floating-point number in IEEE single-precision format to a
+ * 8-bit floating-point number in fp8 E4M3FNUZ format, in bit representation.
+ */
+C10_HOST_DEVICE inline uint8_t fp8e4m3fnuz_from_fp32_value(float f) {
+  /*
+   * Binary representation of 256.0f, which is the first value not representable
+   * (i.e. the first value which would overflow in to the sign bit, resulting in
+   * a NaN) in fp8e4m3fnuz range:
+   * 1 0000 000 - fp8e4m3fnuz
+   * 0 10000110 00000000000000000000000 - fp32
```
Comment (umangyadav) [inline c10/util/Float8_e4m3fnuz.h:75]:
```suggestion
   * 0 10000111 00000000000000000000000 - fp32
```

File: c10/util/Float8_e4m3fnuz.h [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+  const uint32_t sign = f_bits & UINT32_C(0x80000000);
+
+  /*
+   * Set sign bit to 0
+   */
+  f_bits ^= sign;
+
+  if (f_bits >= fnuz_max) {
+    // NaN -- sign bit set to 1, rest 0s
+    return 0x80;
```
Comment (umangyadav) [inline c10/util/Float8_e4m3fnuz.h:107]: table here clips more than FNUZ_MAX to FLT_MAX. 
https://onnx.ai/onnx/technical/float8.html#cast 

Is there reason behind using NaNs ?

File: c10/util/Float8_e4m3fnuz.h [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  if (f_bits < (UINT32_C(121) << 23)) {
```
Comment (umangyadav) [inline c10/util/Float8_e4m3fnuz.h:110]: where this 121 came from ?

File: c10/util/Float8_e4m3fnuz.h [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+  // TORCH_INTERNAL_ASSERT is not supported on a device
+  assert(false);
+#else
+  TORCH_INTERNAL_ASSERT(false, "e4m3fnuz is not supported by CUDA or HIP");
+#endif
+}
+#else
+C10_API float fp8e4m3fnuz_to_fp32_value(uint8_t input);
+#endif
+
```
Comment (alugorey) [inline c10/util/Float8_e4m3fnuz.h:64]: This entire section can be simplified to:
```
#if defined(__CUDA_ARCH__) || defined(__HIP__)
C10_HOST_DEVICE C10_API inline float fp8e4m3fnuz_to_fp32_value(uint8_t) {
  CUDA_KERNEL_ASSERT(false);
}
#else
C10_API float fp8e4m3fnuz_to_fp32_value(uint8_t input);
#endif
```
We can't have a naked `assert` in device code so needs to change to `CUDA_KERNEL_ASSERT`. This should get past your ROCM CI failure. change also needs to happen in e5m4fnuz variant

File: c10/util/Float8_e4m3fnuz.h [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+#include <math.h>
+#include <stdint.h>
+#endif
+
+#include <iosfwd>
+#include <ostream>
+
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+#include <cassert>
+#endif
```
Comment (alugorey) [inline c10/util/Float8_e4m3fnuz.h:39]: This can be removed. See comment below.

**Review by alugorey — CHANGES_REQUESTED** [review]
A couple comments to help with your rocm CI failure. Hope this helps! :)

File: c10/util/Float8_e4m3fnuz.h [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+ * @note The implementation doesn't use any floating-point operations.
+ */
+#if defined(__CUDA__) || defined(__HIP__)
+C10_HOST_DEVICE C10_API inline float fp8e4m3fnuz_to_fp32_value(uint8_t) {
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+  // TORCH_INTERNAL_ASSERT is not supported on a device
+  assert(false);
+#else
+  TORCH_INTERNAL_ASSERT(false, "e4m3fnuz is not supported by CUDA or HIP");
+#endif
```
Comment (malfet) [inline c10/util/Float8_e4m3fnuz.h:59]:
```suggestion
  CUDA_KERNEL_ASSERT(false && "e4m3fnuz is not supported by CUDA or HIP");
  return -1.0;
#endif
```

File: c10/util/Float8_e5m2fnuz.h [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+#include <stdint.h>
+#endif
+
+#include <iosfwd>
+#include <ostream>
+
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+#include <cassert>
+#endif
+
```
Comment (malfet) [inline c10/util/Float8_e5m2fnuz.h:40]:
```suggestion
```

File: c10/util/Float8_e4m3fnuz.h [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+#include <stdint.h>
+#endif
+
+#include <iosfwd>
+#include <ostream>
+
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+#include <cassert>
+#endif
+
```
Comment (malfet) [inline c10/util/Float8_e4m3fnuz.h:40]:
```suggestion
```

File: c10/util/Float8_e5m2fnuz.h [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+ * @note The implementation doesn't use any floating-point operations.
+ */
+#if defined(__CUDA__) || defined(__HIP__)
+C10_HOST_DEVICE C10_API inline float fp8e5m2fnuz_to_fp32_value(uint8_t) {
+#if defined(__CUDA_ARCH__) || defined(__HIP_ARCH__)
+  // TORCH_INTERNAL_ASSERT is not supported on a device
+  assert(false);
+#else
+  TORCH_INTERNAL_ASSERT(false, "e5m2fnuz is not supported by CUDA or HIP");
+#endif
```
Comment (malfet) [inline c10/util/Float8_e5m2fnuz.h:59]:
```suggestion
  CUDA_KERNEL_ASSERT(false && "e5m2fnuz is not supported by CUDA or HIP");
  return -1.0
```

File: c10/util/Float8_e4m3fnuz.h (hunk 12 above)
Comment (galexite) [reply c10/util/Float8_e4m3fnuz.h:107]: The reason is that the existing casting code for `e5m2` and `e4m3fn` types are also implemented without any saturation, so I chose to do the same to match that behaviour.

File: c10/util/Float8_e4m3fnuz.h (hunk 13 above)
Comment (galexite) [reply c10/util/Float8_e4m3fnuz.h]: Fixed and comment added.

File: c10/util/Float8_e4m3fnuz.h (hunk 12 above)
Comment (umangyadav) [reply c10/util/Float8_e4m3fnuz.h:99]: I fear it would lead to having more NaNs when doing inference.  Ideally there should be a flag for the users to set which behaviour they want
